Added did:object

mwherman2000 opened the pull request:
----- DID METHOD REGISTRATION FORM: DELETE EVERYTHING ABOVE THIS LINE ------

### DID Method Registration

As a DID method registrant, I have ensured that my DID method registration complies with the following statements:

- [X] The DID Method specification [defines the DID Method Syntax](https://w3c.github.io/did-core/#method-syntax).
- [X] The DID Method specification [defines the Create, Read, Update, and Deactivate DID Method Operations](https://w3c.github.io/did-core/#method-operations).
- [X] The DID Method specification [contains a Security Considerations section](https://w3c.github.io/did-core/#security-requirements).
- [X] The DID Method specification [contains a Privacy Considerations section](https://w3c.github.io/did-core/#privacy-requirements).
- [X] The JSON file I am submitting has [passed all automated validation tests below](#partial-pull-merging).
- [X] The JSON file contains a `contactEmail` address [OPTIONAL].
- [X] The JSON file contains a `verifiableDataRegistry` entry [OPTIONAL].

mprorock commented:
@mwherman2000 if i am interpreting correctly this method appears pretty developer oriented to be used as part of a broader toolchain.  If so, might be worth pointing from the method spec in the appropriate sections to the security and privacy section of the [did impl guide](https://w3c.github.io/did-imp-guide/#privacy-considerations) for developers so they get some pointers at best practices.   thoughts?

mprorock reviewed (changes requested):
@mwherman2000 I would suggest that for key types to leverage existing types/approaches - see [did:key](https://w3c-ccg.github.io/did-method-key/#create) for an example, or to utilize JWK to ease adoption

inline suggestion here:

```
{
    "id": "did:object:AGsL32ZMvAwxYRN9Sv4mrgu3DgBSvTm5vt",
    "publicKey": [{
        "id": "did:object:AGsL32ZMvAwxYRN9Sv4mrgu3DgBSvTm5vt#keys-1",
        "type": ["ECDSA", "secp256r1"],
        "publicKeyHex": "02ba8ba8e8ef1f313ecc095ba0bf0d1a921a138346d6817812714f4a9e4cca8ccf"
    },
```

adjust to:
```
{
    "id": "did:object:AGsL32ZMvAwxYRN9Sv4mrgu3DgBSvTm5vt",
    "publicKey": [{
        "id": "did:object:AGsL32ZMvAwxYRN9Sv4mrgu3DgBSvTm5vt#keys-1",
        "type": ["Ed25519VerificationKey2018"],
        "publicKeyHex": "02ba8ba8e8ef1f313ecc095ba0bf0d1a921a138346d6817812714f4a9e4cca8ccf"
    },
```

hard change requested is to get the pub key type in line, potential use of JWK is just a suggestion

msporny reviewed (changes requested):
It's difficult to believe that the DID Method does not have any security or privacy considerations. For example, you can switch between PoW, PoS, and PoA on the Stratis Platform ... which means that you have different security characteristics that implementers should be aware of in the Security Considerations section.

Please reconsider the contents of the Security and Privacy considerations section. Other than that, the DID Method meets the minimum requirements for provisional registration.

mwherman2000 commented:
> It's difficult to believe that the DID Method does not have any security or privacy considerations. For example, you can switch between PoW, PoS, and PoA on the Stratis Platform ... which means that you have different security characteristics that implementers should be aware of in the Security Considerations section.
> 
> Please reconsider the contents of the Security and Privacy considerations section. Other than that, the DID Method meets the mini

mwherman2000 commented:
> It's difficult to believe that the DID Method does not have any security or privacy considerations. For example, you can switch between PoW, PoS, and PoA on the Stratis Platform ... which means that you have different security characteristics that implementers should be aware of in the Security Considerations section.
> 
> Please reconsider the contents of the Security and Privacy considerations section. Other than that, the DID Method meets the minimum requirements for provisional registration.

First, my apologies, I'm replying from my phone. 

This method doesn't have any inherent security or privacy considerations itself. It's only a method. How it is used ...how it is implemented ...whether it's used to secure kiss NFTs or SWIFT transactions all make a difference ...it would take multiple college courses to cover all of this.

What is the best example text for these sections that the approvers have seen in other method specifications?

Note: the did:ont specification from which this spec is based has no content in these sections. The only technical requirement is to have a section called Security Considerations and Privacy Considerations. There needs to be a level playing field here.

mwherman2000 commented:
> @mwherman2000 if i am interpreting correctly this method appears pretty developer oriented to be used as part of a broader toolchain. If so, might be worth pointing from the method spec in the appropriate sections to the security and privacy section of the [did impl guide](https://w3c.github.io/did-imp-guide/#privacy-considerations) for developers so they get some pointers at best practices. thoughts?

I can add those pointers. No problem. Is this sufficient for you as well @msporny ?

mwherman2000 commented:
> @mwherman2000 I would suggest that for key types to leverage existing types/approaches - see [did:key](https://w3c-ccg.github.io/did-method-key/#create) for an example, or to utilize JWK to ease adoption

I'm willing to make the change but don't understand what is being asked for.

Also thye did:key example uses publicKeyBase58 ...which is deprecated, n'est pas?

msporny commented:
>  Is this sufficient for you as well @msporny ?

Not really, though I might be missing the point of `did:object`? Security and Privacy considerations sections that are largely empty tend to be fairly suspect. Looking into the ledger mechanism that's being used for `did:object` raised security/privacy questions (as I alluded to in my previous comment). If `did:object` is a "meta DID Method", then that should probably be stated clearly... but given that its tied to DLTs of some form (with three consensus methods to pick from)... it feels like there are trade-offs that are being made that need to be documented.

> What is the best example text for these sections that the approvers have seen in other method specifications?

Here are a few (none of them are perfect, but they demonstrate a good understanding of the sorts of considerations we expect to see): 

https://did-tezos.spruceid.com/#security-privacy
https://w3c-ccg.github.io/did-method-key/#security-and-privacy-considerations
https://identity.foundation/peer-did-method-spec/index.html#security-considerations
https://w3c-ccg.github.io/did-method-web/#security-and-privacy-considerations
https://identity.foundation/keri/did_methods/#privacy-considerations
https://github.com/trustbloc/trustbloc-did-method/blob/main/docs/spec/trustbloc-did-method.md#security-considerations

Again, since you're doing a provisional registration, you don't have to be exhaustive.

mwherman2000 commented:
@mprorock I've updated https://github.com/mwherman2000/TrustedDigitalWeb/blob/master/specifications/did-methods/did-object.md#example-2-trusted-digital-web-did-object-did-document per your request.

@msporny and @mprorock I've updated the Security and Privacy Considerations sections per your initial feedback: https://github.com/mwherman2000/TrustedDigitalWeb/blob/master/specifications/did-methods/did-object.md#4-security-considerations

@msporny I will also look through your additional examples this evening.  "More news at 11..."

I believe `did:object` is going to be quite an important development in the long run and very much appreciate your support and advice.

mwherman2000 commented:
@msporny @mprorock Security and Privacy sections: added Manu's list of best-in-class methods (for now); https://github.com/mwherman2000/TrustedDigitalWeb/blob/master/specifications/did-methods/did-object.md#4-security-considerations

I also added an Acknowledgements section: https://github.com/mwherman2000/TrustedDigitalWeb/blob/master/specifications/did-methods/did-object.md#7-acknowledgments

mwherman2000 reviewed (commented):
Changes were completed per feedback in the comments: https://github.com/mwherman2000/TrustedDigitalWeb/blob/master/specifications/did-methods/did-object.md#4-security-considerations

mwherman2000 reviewed (commented):
Changes were made per feedback in the comments: https://github.com/mwherman2000/TrustedDigitalWeb/blob/master/specifications/did-methods/did-object.md#example-2-trusted-digital-web-did-object-did-document

mwherman2000 reviewed (commented):
I believe this request has been satisfied as well.

mwherman2000 commented:
> > Is this sufficient for you as well @msporny ?
> 
> Not really, though I might be missing the point of `did:object`? Security and Privacy considerations sections that are largely empty tend to be fairly suspect. Looking into the ledger mechanism that's being used for `did:object` raised security/privacy questions (as I alluded to in my previous comment). If `did:object` is a "meta DID Method", then that should probably be stated clearly... but given that its tied to DLTs of some form (with three consensus methods to pick from)... it feels like there are trade-offs that are being made that need to be documented.
> 
> > What is the best example text for these sections that the approvers have seen in other method specifications?
> 
> Here are a few (none of them are perfect, but they demonstrate a good understanding of the sorts of considerations we expect to see):
> 
> https://did-tezos.spruceid.com/#security-privacy https://w3c-ccg.github.io/did-method-key/#security-and-privacy-considerations https://identity.foundation/peer-did-method-spec/index.html#security-considerations https://w3c-ccg.github.io/did-method-web/#security-and-privacy-considerations https://identity.foundation/keri/did_methods/#privacy-considerations https://github.com/trustbloc/trustbloc-did-method/blob/main/docs/spec/trustbloc-did-method.md#security-considerations
> 
> Again, since you're doing a provisional registration, you don't have to be exhaustive.

While the Stratis Platform *code repository* technically supports 3 consensus schemes only 1 is in effect for the entire blockchain ...chosen during initial build and deployment of the global Trusted Digital Web VDR (e.g. on the Mainnet). The consensus issue is a bit of a red herring. App, agent, and service implementors don't get to chose the consensus scheme.

kdenhartog commented:
FYI making a verification method an Array type is not compliant with the DID Document data model. See [section 5.2](https://w3c.github.io/did-core/#verification-methods) where it specifies the `type` property MUST be a string.

mwherman2000 commented:
> FYI making a verification method an Array type is not compliant with the DID Document data model. See [section 5.2](https://w3c.github.io/did-core/#verification-methods) where it specifies the `type` property MUST be a string. I believe that's what @mprorock was referring to above.

@kdenhartog Is @mprorock 's example correct? ...to me it still appears to use an array...
```json
{
    "id": "did:object:AGsL32ZMvAwxYRN9Sv4mrgu3DgBSvTm5vt",
    "publicKey": [{
        "id": "did:object:AGsL32ZMvAwxYRN9Sv4mrgu3DgBSvTm5vt#keys-1",
        "type": ["Ed25519VerificationKey2018"]...
```

> I'll have to spend more time giving a more thorough review when I've got more time to read through this.

Hold off on any further reviews.  There's some fundamental issues (objections) with the assessment process and I'm going to be cancelling this PR to open a new PR with a significantly refactored DID Method specification for "DID Objects". "More news in a few minutes..."

mwherman2000 commented:
Closed due to issue: https://github.com/w3c/did-spec-registries/issues/387

A new PR for version 1.5.0 will be opened shortly.
